bindings/rust/build.rs

```diff
@@ -1,9 +1,122 @@
+// is not used as a depedency to avoid double link issues on dependants.
+fn compile_blst(blst_base_dir: PathBuf) {
+    // account for cross-compilation [by examining environment variables]
+    let target_os = env::var("CARGO_CFG_TARGET_OS").unwrap();
+    let target_arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap();
+
+    if target_os.ne("none") && !env::var("BLST_TEST_NO_STD").is_ok() {
+        println!("cargo:rustc-cfg=feature=\"std\"");
+        if target_arch.eq("wasm32") {
+            println!("cargo:rustc-cfg=feature=\"no-threads\"");
+        }
+    }
+    println!("cargo:rerun-if-env-changed=BLST_TEST_NO_STD");
+
+    println!("Using blst source directory {}", blst_base_dir.display());
+
+    // Set CC environment variable to choose alternative C compiler.
+    // Optimization level depends on whether or not --release is passed
+    // or implied.
+
+    #[cfg(target_env = "msvc")]
+    if env::var("CARGO_CFG_TARGET_POINTER_WIDTH").unwrap().eq("32") && !env::var("CC").is_ok() {
+        match std::process::Command::new("clang-cl")
+            .arg("--version")
+            .output()
+        {
+            Ok(out) => {
+                if String::from_utf8(out.stdout)
+                    .unwrap_or("unintelligible".to_string())
+                    .contains("Target: i686-")
+                {
+                    env::set_var("CC", "clang-cl");
+                }
+            }
+            Err(_) => { /* no clang-cl in sight, just ignore the error */ }
+        }
+    }
+
+    let mut cc = cc::Build::new();
+
+    let c_src_dir = blst_base_dir.join("src");
+    println!("cargo:rerun-if-changed={}", c_src_dir.display());
+    let mut file_vec = vec![c_src_dir.join("server.c")];
+
+    if target_arch.eq("x86_64") || target_arch.eq("aarch64") {
+        let asm_dir = blst_base_dir.join("build");
+        println!("cargo:rerun-if-changed={}", asm_dir.display());
+        blst_assembly(&mut file_vec, &asm_dir, &target_arch);
+    } else {
+        cc.define("__BLST_NO_ASM__", None);
+    }
+    match (cfg!(feature = "portable"), cfg!(feature = "force-adx")) {
+        (true, false) => {
+            println!("Compiling in portable mode without ISA extensions");
+            cc.define("__BLST_PORTABLE__", None);
+        }
+        (false, true) => {
+            if target_arch.eq("x86_64") {
+                println!("Enabling ADX support via `force-adx` feature");
+                cc.define("__ADX__", None);
+            } else {
+                println!("`force-adx` is ignored for non-x86_64 targets");
+            }
+        }
+        (false, false) =>
+        {
+            #[cfg(target_arch = "x86_64")]
+            if target_arch.eq("x86_64") && std::is_x86_feature_detected!("adx") {
+                println!("Enabling ADX because it was detected on the host");
+                cc.define("__ADX__", None);
+            }
+        }
```

Comment:
I think it would be nice if there was a way to force `__SHA__` and `__ARM_FEATURE_CRYPTO` too but I'm not sure it's worth the complexity. Unless you feel like it should be there too, don't worry about it. For reference, [this is where this is checked in blst](https://github.com/supranational/blst/blob/a7fd1f584d26b0ae6cdc427976ea1d8980f7e15d/src/sha256.h#L14-L15).